More robust interceptors in case of exception

whut opened the pull request:
Hey,

Recently I discovered in my project that when one interceptor throws in IRequestHandlerInterceptor.BeforeHandlingRequest, for all previously executed interceptors IRequestHandlerInterceptor.AfterHandlingRequest is not executed. In our project it would kill application because of some work that really need to be done after handling every request, but not when BeforeHandlingRequest was not called (so using Dispose is not clean solution).

I believe that Agatha should guarantee that for every interceptor that had BeforeHandlingRequest called, the AfterHandlingRequest  also will be. ASP.NET MVC action filters behave in such way for example.

Commit:
Execute IRequestHandlerInterceptor.AfterHandlingRequest for all interceptors that had BeforeHandlingRequest executed, even when some other BeforeHandlingRequest throwed exception

To make it happened moved handling exception in request handler to IRequestProcessingErrorHandler

Commit:
Execute IRequestHandlerInterceptor.AfterHandlingRequest for all interceptors that had BeforeHandlingRequest executed, even when some other AfterHandlingRequest throwed exception

Now Agatha guarantees that when BeforeHandlingRequest gets called, it's AfterHandlingRequest also will be, even in case of exceptions in other interceptors.

Commit:
Added null check when releasing interceptors

When resolving interceptors fails, there are no interceptors to release.

whut commented:
Hi,

Could you review and hopefully merge my pull request?

whut commented:
I rebased and updated commits against newest changes in Agatha

davybrion commented:
it looks good to me, but i'm gonna wait until i get feedback from @bartdeleye on this

bartdeleye commented:
We generally solve the issue you describe by setting a flag during BeforeHandlingRequest() and/or AfterHandlingRequest() so that we can implement the correct behavior during Dispose().
To deal with exceptions in your interceptor you can put a try-catch block in the BeforeHandlingRequest() body and mark the request as processed with a response including the exception. The RequestHandler won't be invoked and AfterHandlingRequest() will be executed for every invoked interceptor. Again to make your interceptors robust, you can put a try-catch block in the body of AfterHandlingRequest() to make sure that all invoked interceptors will have AfterHandlingRequest() executed.

Anyway, I reviewed your changes in case the solution above doesn't fullfil your needs, but noticed some issues:
-in case a request handler throws an exception and an interceptor throws one too during AfterHandlingRequest(), the original exception will be overwritten which hides the root cause of the problem. The same goes for an exception that occurred during BeforeHandlingRequest().
-before, any exception that occured outside of the request handler, would result in a 500 (rethrown). Maybe this could "break" some existing client code.
-make sure that dispose is always called, even if the error handler throws an exception.

whut committed:
Fixed memory leak when resolving request handler, but not releasing it

whut commented:
Yes, I can work around this, but I think it would be good to have Agatha handling that.

About overriding request handler exception with interceptor one, it is like this also in ASP.NET MVC, and I would say that it is OK: Agatha tries it best to handle exception and return them using Response.Exception, but in case you mess something in interceptor, it is your fault:)

Second, yep, it is change in behavior, but I think it's a good change.

Third, I fixed that, sorry to break it in the first place.

I also rebased changes, and added another commit with added Container.Release after Container.Resolve in error handler.

whut committed:
Added tests for request interceptor handling in case of exception

davybrion commented:
i'm gonna merge this, but i'm gonna add a commit later on that replaces the lastExceptionFromInterceptor with an IEnumerable with every thrown exception by the interceptor.

For one, I believe they all should be logged and secondly, i think errorHandler.DealWithException should be called with the _first_ exception, since that is more likely to be the root cause of the problem.  The last thrown exception in most cases will only be a consequence of the first exception.

davybrion commented:
PS: sorry for the multiple delays in dealing with this pull request... won't happen again :)

whut commented:
Thanks, it is open source, I don't mind delays.

About handling exceptions, I closely followed ASP.NET MVC behavior, even without considering it:)
